applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -0,0 +1,856 @@
+    {
+        const array_1d<double, 3>& coords = GetGeometry()[i].Coordinates();
+        for (unsigned int k = 0; k < Dim; ++k)
+        {
+            Points(i, k) = coords[k];
+        }
+    }
+
+    const unsigned int nsubdivisions = EnrichmentUtilities::CalculateEnrichedShapeFuncions(
+        Points, data.DN_DX, data.distances, Volumes, GPShapeFunctionValues,
+        PartitionsSign, GradientsValue, NEnriched);
+
+    const double density = ComputeDensity(rCurrentProcessInfo);
+    const double DrhoDu2 = ComputeDensityDerivative(density, rCurrentProcessInfo);
+
+    // Computing local velocity
+    const array_1d<double, 3> free_stream_velocity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+    array_1d<double, Dim> velocity = PotentialFlowUtilities::ComputeVelocity<Dim,NumNodes>(*this);
+    for (unsigned int i = 0; i < Dim; i++){
+        velocity[i] += free_stream_velocity[i];
+    }
+
+    const BoundedVector<double, NumNodes> DNV = prod(data.DN_DX, velocity);
+
+    // Compute the lhs and rhs that would correspond to it being divided
+    for (unsigned int i = 0; i < nsubdivisions; ++i)
+    {
+        if (PartitionsSign[i] > 0)
+        {
+            noalias(lhs_positive) +=
+                Volumes[i] * density * prod(data.DN_DX, trans(data.DN_DX));
+            noalias(lhs_positive) +=
+                Volumes[i] * 2 * DrhoDu2 * outer_prod(DNV, trans(DNV));
+        }
+        else
+        {
+            noalias(lhs_negative) +=
+                Volumes[i] * density * prod(data.DN_DX, trans(data.DN_DX));
+            noalias(lhs_negative) +=
+                Volumes[i] * 2 * DrhoDu2 * outer_prod(DNV, trans(DNV));
+        }
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::CalculateVolumesSubdividedElement(
+    double& rUpper_vol,
+    double& rLower_vol,
+    const ProcessInfo& rCurrentProcessInfo)
+{
+    ElementalData<NumNodes, Dim> data;
+
+    // Calculate shape functions
+    GeometryUtils::CalculateGeometryData(GetGeometry(), data.DN_DX, data.N, data.vol);
+
+    GetWakeDistances(data.distances);
+
+    // Subdivide the element
+    constexpr unsigned int nvolumes = 3 * (Dim - 1);
+    BoundedMatrix<double, NumNodes, Dim> Points;
+    array_1d<double, nvolumes> PartitionsSign;
+    BoundedMatrix<double, nvolumes, NumNodes> GPShapeFunctionValues;
+    array_1d<double, nvolumes> Volumes;
+    std::vector<Matrix> GradientsValue(nvolumes);
+    BoundedMatrix<double, nvolumes, 2> NEnriched;
+    for (unsigned int i = 0; i < GradientsValue.size(); ++i)
+        GradientsValue[i].resize(2, Dim, false);
+    for (unsigned int i = 0; i < NumNodes; ++i)
+    {
+        const array_1d<double, 3>& coords = GetGeometry()[i].Coordinates();
+        for (unsigned int k = 0; k < Dim; ++k)
+        {
+            Points(i, k) = coords[k];
+        }
+    }
+
+    const unsigned int nsubdivisions = EnrichmentUtilities::CalculateEnrichedShapeFuncions(
```

Comment:
Enrichment utilities is deprecated. Besides this you are computing (and allocating) quite a lot of things to get the volumes.

Please use the `ModifiedShapeFunctions` classes. These are implemented in the KratosCore and extensively used in the FluidDynamicsApplication.

Reply:
@eloisabaez I will do this in another PR

Reply:
This is a new feature PR. Deprecated stuff should never be included in new features.

In any case as far as I see you only require the volumes of the subelements. I think that in this context it might have sense to enhance the interface of either the `ModifiedShapeFunctions` or the `DivideGeometry` utils to just compute this an avoid the shape functions and gradients overhead.

Reply:
Makes sense, Ill make this change in all of the elements once this is available.

Reply:
OK then. I'll do it ASAP.

Reply:
Thanks!